### PR TITLE
Smoke test for parallel-netcdf package

### PR DESCRIPTION
**Pull request by sshudler**
This PR represents a preliminary smoke test for the `parallel-netcdf` package and is intended to serve the following purposes:

1. Leverage the current E4S test suite for the software; and
2. Demonstrate to package maintainers how to start a Spack smoke test.

### CODE REVIEW

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py [hunk 1]
```diff
@@ -169,3 +171,34 @@ def configure_args(self):
...
+
+    @run_after('install')
+    def cache_test_sources(self):
+        """Copy the example source files after the package is installed to an
+        install test subdirectory for use during `spack test run`."""
+        self.cache_extra_test_sources([self.examples_src_dir])
+
+    def test(self):
+        test_dir = join_path(self.install_test_root, self.examples_src_dir)
+        with working_dir(test_dir, create=False):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/parallel-netcdf/package.py]: Performing the test using `with working_dir` is not needed when you're using `self.run_test`.  You should only need to supply the `work_dir` argument.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
LGTM though I would prefer use of the `work_dir` argument versus use of `with working_dir`.  Is there a reason you chose to go with the latter?

**Comment by tldahlgren** [comment]
@skosukhin Care to provide review feedback on a preliminary/example smoke test for the package?

**Comment by skosukhin** [comment]
I am not really familiar with this functionality but I have two observations:
1. Spack seems to set environment variables (e.g. `LIBRARY_PATH`, `C_INCLUDE_PATH`, `LD_LIBRARY_PATH`) that tell the compiler (and the dynamic linker) where the headers and libraries of the tested package can be found: https://github.com/spack/spack/blob/df0cc9ee028b27fc136a72bebc3d5665ce8371a9/lib/spack/spack/build_environment.py#L757 If we rely on that, we don't need to set `-I` and `-L` flags. Otherwise, we should also add `rpath` flags.
2. The package itself uses this logic to detect the MPI run command to be used for testing: https://github.com/Parallel-NetCDF/PnetCDF/blob/70db79959aaefaa8c6a93b8ec7c70219ed4c2d8d/configure.ac#L2381-L2389. It might make sense to make sure that the same command is used in both places. Also, note that `mpirun` might be intentionally broken: https://github.com/spack/spack/blob/df0cc9ee028b27fc136a72bebc3d5665ce8371a9/var/spack/repos/builtin/packages/openmpi/package.py#L834-L857

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py (hunk 1 above)
Comment (sshudler) [reply var/spack/repos/builtin/packages/parallel-netcdf/package.py]: @tldahlgren That's a good point. Perhaps `with working_dir` makes more sense when I actually need to create a temporary directory. I'll try setting only the `work_dir` argument, if it works I'll remove the `with working_dir`

**Comment by sshudler** [comment]
> I am not really familiar with this functionality but I have two observations:
> 
>     1. Spack seems to set environment variables (e.g. `LIBRARY_PATH`, `C_INCLUDE_PATH`, `LD_LIBRARY_PATH`) that tell the compiler (and the dynamic linker) where the headers and libraries of the tested package can be found: https://github.com/spack/spack/blob/df0cc9ee028b27fc136a72bebc3d5665ce8371a9/lib/spack/spack/build_environment.py#L757
>         If we rely on that, we don't need to set `-I` and `-L` flags. Otherwise, we should also add `rpath` flags.
> 

@skosukhin That's a good point - it works without `-I` and `-L`

>     2. The package itself uses this logic to detect the MPI run command to be used for testing: https://github.com/Parallel-NetCDF/PnetCDF/blob/70db79959aaefaa8c6a93b8ec7c70219ed4c2d8d/configure.ac#L2381-L2389. It might make sense to make sure that the same command is used in both places. Also, note that `mpirun` might be intentionally broken: https://github.com/spack/spack/blob/df0cc9ee028b27fc136a72bebc3d5665ce8371a9/var/spack/repos/builtin/packages/openmpi/package.py#L834-L857

I see what you mean. In the smoke tests of `openmpi` package they actually check for `mpirun` but set `skip_missing=True` which means it will silently fail if `mpirun` is missing. My solution would be just to pre-check if any of the variants `['mpirun', 'mpiexec', 'srun']` exists using `os.path.isfile` and then use the first that exists.

**Review by tldahlgren — APPROVED** [review]
LGTM as a preliminary smoke test.

**Comment by tldahlgren** [comment]
Thank you for your contribution!

**Pull request merged**